### PR TITLE
Feat/geometry lod

### LINKED ISSUE #155: Task - Generate simplified geometry on-the-fly for returned objets - impl
Implementation phase following #143

### PR DESCRIPTION
**Pull request by nick-rv**

### CODE REVIEW

File: internal/data/feature_mock.go [hunk 1]
```diff
@@ -135,6 +161,11 @@ func MakeFeatureMockPointAsJSON(tableName string, id int, x float64, y float64,
 	return feat.toJSON(columns)
 }
 
+func MakeFeatureMockPolygonAsJSON(tableName string, id int, coords orb.Ring, columns []string) string {
```
Comment (benoitdm-oslandia) [inline internal/data/feature_mock.go:166]: add comment to public function

File: internal/data/feature_mock.go [hunk 2]
```diff
@@ -156,3 +187,16 @@ func MakeFeaturesMockPoint(tableName string, extent api.Extent, nx int, ny int)
 	}
 	return features
 }
+
+// Returns a slice of polygon featureMocks with as many entries into the coords slice as argument
+// val is an arbitraty value which is used to populate the properties values from each Feature
+func MakeFeaturesMockPolygon(tableName string, val int, coords []orb.Ring) []*featureMock {
```
Comment (benoitdm-oslandia) [inline internal/data/feature_mock.go]: replace `val` by `id`

File: internal/data/feature_mock.go [hunk 3]
```diff
@@ -156,3 +187,16 @@ func MakeFeaturesMockPoint(tableName string, extent api.Extent, nx int, ny int)
...
+
+	features := make([]*featureMock, len(coords))
+	index := 0
+	for _, coord := range coords {
+		features[index] = makeFeatureMockPolygon(tableName, val, coord)
+		index++
```
Comment (benoitdm-oslandia) [inline internal/data/feature_mock.go]: you do not need to use `index` you can use `features = append(features, ...)`

File: internal/service/db_test/handler_db_lod_test.go [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+	"github.com/CrunchyData/pg_featureserv/internal/api"
+	util "github.com/CrunchyData/pg_featureserv/internal/utiltest"
+	"github.com/paulmach/orb"
+	"github.com/paulmach/orb/geojson"
+)
+
+// Simple unit test case ensuring that simplification is working on a single feature
+func (t *DbTests) TestGeometrySimplificationSingleFeature() {
+	t.Test.Run("TestGeometrySimplificationSingleFeature", func(t *testing.T) {
+		rr := hTest.DoRequest(t, "/collections/mock_poly/items/6?max-allowable-offset=0.01")
```
Comment (benoitdm-oslandia) [inline internal/service/db_test/handler_db_lod_test.go]: can you first check this item has (without simplification) a 'big' geometry?

File: internal/service/db_test/handler_db_lod_test.go [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+func (t *DbTests) TestGeometrySimplificationSeveralFeatures() {
+	t.Test.Run("TestGeometrySimplificationSeveralFeatures", func(t *testing.T) {
+		rr := hTest.DoRequest(t, "/collections/mock_poly/items?max-allowable-offset=0.01")
+		// Feature collection
+		var v api.FeatureCollection
+		errUnMarsh := json.Unmarshal(hTest.ReadBody(rr), &v)
+		util.Assert(t, errUnMarsh == nil, fmt.Sprintf("%v", errUnMarsh))
+		util.Equals(t, 6, len(v.Features), "wrong number of features")
+		feature := v.Features[0]
+		util.Equals(t, 4, len(feature.Geom.Geometry().(orb.Polygon)[0]), "wrong number of simplified coordinates")
```
Comment (benoitdm-oslandia) [inline internal/service/db_test/handler_db_lod_test.go]: can you just check the `len(v.Features) > 1` and then use a for loop to check all the feature of `v.Features`?

File: internal/service/db_test/handler_db_lod_test.go [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+	})
+}
+
+// Test case with negative value as simplification factor
+func (t *DbTests) TestGeometrySimplificationNegativeValue() {
+	t.Test.Run("TestGeometrySimplificationNegativeValue", func(t *testing.T) {
+		path := "/collections/mock_poly/items/6?max-allowable-offset=-0.01"
+		// If lower thant minVal, then minValue (0) is considered
```
Comment (benoitdm-oslandia) [inline internal/service/db_test/handler_db_lod_test.go]: typo: `than` not `thant`

Also, I think this choice must be discussed with the CrunchyData team. I tend to prefer an error like the `http.StatusBadRequest`

File: internal/service/db_test/handler_db_lod_test.go [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+		util.Assert(t, errUnMarsh == nil, fmt.Sprintf("%v", errUnMarsh))
+		util.Equals(t, 4, len(feat.Geom.Geometry().(orb.Polygon)[0]), "wrong number of simplified coordinates")
+
+		path = "/collections/mock_poly/items/4?max-allowable-offset=0.001"
+		rr = hTest.DoRequestMethodStatus(t, "GET", path, nil, nil, http.StatusOK)
+		errUnMarsh = json.Unmarshal(hTest.ReadBody(rr), &feat)
+		util.Assert(t, errUnMarsh == nil, fmt.Sprintf("%v", errUnMarsh))
+		util.Equals(t, 10, len(feat.Geom.Geometry().(orb.Polygon)[0]), "wrong number of simplified coordinates")
+
+		path = "/collections/mock_poly/items/4?max-allowable-offset=1"
```
Comment (benoitdm-oslandia) [inline internal/service/db_test/handler_db_lod_test.go]: May be returning an empty geometry should be considered as something not acceptable and should raise an error. To check with CrunchyData

File: internal/service/handler.go [hunk 8]
```diff
@@ -465,7 +465,6 @@ func handleCollectionItems(w http.ResponseWriter, r *http.Request) *appError {
 	}
 	param, errQuery := createQueryParams(&reqParam, tbl.Columns, tbl.Srid)
 	param.Filter = parseFilter(reqParam.Values, tbl.DbTypes)
-
```
Comment (benoitdm-oslandia) [inline internal/service/handler.go:468]: revert this change please

File: internal/service/param.go [hunk 9]
```diff
@@ -145,6 +146,13 @@ func parseRequestParams(r *http.Request) (RequestParam, error) {
 		return param, err
 	}
 
+	// --- max-allowable-offset parameter
+	// parseFloat(paramValues, key string, minVal float64, maxVal float64, defaultVal float64)
+	param.MaxAllowableOffset, err = parseFloat(paramValues, api.ParamMaxAllowableOffset, 0, 100000, 0)
```
Comment (benoitdm-oslandia) [inline internal/service/param.go:151]: do you have an explanation for the 100000 value?

File: internal/utiltest/test_db.go [hunk 10]
```diff
@@ -114,7 +114,18 @@ func InsertSimpleDataset(db *pgxpool.Pool, schema string) {
...
+		CloseTestDb(db)
+		log.Fatal(errExec)
+	}
+
+	// ================================================================================
+
+	// inserting point features into mock collections
 	b := &pgx.Batch{}
 
 	insertBytes := []byte(`
```
Comment (benoitdm-oslandia) [inline internal/utiltest/test_db.go]: As sonarcloud suggests: we should extract this part in a dedicated function

File: internal/utiltest/test_db.go [hunk 11]
```diff
@@ -141,6 +152,33 @@ func InsertSimpleDataset(db *pgxpool.Pool, schema string) {
...
+		if res == nil {
+			CloseTestDb(db)
+			log.Fatal("Injection failed")
+		}
+		resClose := res.Close()
+		if resClose != nil {
+			CloseTestDb(db)
+			log.Fatal(fmt.Sprintf("Injection failed: %v", resClose.Error()))
+		}
+	}
```
Comment (benoitdm-oslandia) [inline internal/utiltest/test_db.go]: we should extract this part in a dedicated function

**Review by benoitdm-oslandia — CHANGES_REQUESTED** [review]
Still some minor corrections! Good job!

**Comment by benoitdm-oslandia** [comment]
Last thing: can you add some documentation about how using this query parameter? What does it mean when a small value is used or a big one?

File: internal/service/param.go (hunk 9 above)
Comment (nick-rv) [reply internal/service/param.go:151]: It has been chosen in an arbitrary manner.
I tend to consider that using more than 100000 meters or 100000 degrees can be considered as an upper limit.
But it seems i need some more opinions ?

File: internal/api/openapi.go [hunk 12]
```diff
@@ -500,6 +500,23 @@ func GetOpenAPIContent(urlBase string) *openapi3.T {
...
 	}
+	paramMaxAllowableOffset := openapi3.ParameterRef{
+		Value: &openapi3.Parameter{
+			Name:        "max-allowable-offset",
+			Description: "Tolerance to apply for geometry simplification on returned feature(s).",
+			In:          "query",
+			Required:    false,
+			Schema: &openapi3.SchemaRef{
+				Value: &openapi3.Schema{
+					Type:    "integer",
```
Comment (benoitdm-oslandia) [inline internal/api/openapi.go]:
```suggestion
					Type:    "number",
```
ie. float

File: internal/utiltest/test_db.go [hunk 13]
```diff
@@ -78,34 +78,36 @@ func CreateSchema(db *pgxpool.Pool, schema string) {
...
 func InsertSimpleDataset(db *pgxpool.Pool, schema string) {
 	ctx := context.Background()
+
 	// collections tables
-	// tables := []string{"mock_a", "mock_b", "mock_c"}
 	type tableContent struct {
-		extent api.Extent
-		nx     int
-		ny     int
+		geometryType string
```
Comment (benoitdm-oslandia) [inline internal/utiltest/test_db.go]: You have created a `GeometryType` in `datatype.go`, can you use it here?

**Review by benoitdm-oslandia — CHANGES_REQUESTED** [review]
Congrat! Only 4 remaining comments and we will merge!

File: internal/service/db_test/handler_db_lod_test.go (hunk 7 above)
Comment (benoitdm-oslandia) [reply internal/service/db_test/handler_db_lod_test.go]: @pramsey Do you have a preference?

File: internal/service/param.go (hunk 9 above)
Comment (benoitdm-oslandia) [reply internal/service/param.go:151]: ok @nrevelant we keep it like that!

**Review by benoitdm-oslandia — DISMISSED** [review]
We will wait a bit for CrunchyData before merging

**Review by benoitdm-oslandia — APPROVED** [review]

**Review by benoitdm-oslandia — DISMISSED** [review]
LGTM

File: internal/service/db_test/handler_db_lod_test.go (hunk 7 above)
Comment (benoitdm-oslandia) [reply internal/service/db_test/handler_db_lod_test.go]: may be later

**Review by benoitdm-oslandia — APPROVED** [review]

**Pull request merged**